File: backend/game/state_manager.py
import uuid
from game.models import GameState, Player, Country
from game.map_data import COUNTRIES, START_COUNTRIES

# In-memory store: game_id -> GameState
game_sessions: dict[str, GameState] = {}


def _make_countries() -> dict[str, Country]:
    return {name: Country(name=name, **data) for name, data in COUNTRIES.items()}


def create_session(creator_name: str) -> tuple[str, str]:
    """Create a new game session. Returns (game_id, creator_player_id)."""
    game_id = str(uuid.uuid4())
    player_id = str(uuid.uuid4())

    start_country = START_COUNTRIES[0]
    countries = _make_countries()
    countries[start_country].owner_id = player_id

    player = Player(
        player_id=player_id,
        name=creator_name,
        empire=[start_country],
    )

    state = GameState(
        game_id=game_id,
        creator_id=player_id,
        players=[player],
        countries=countries,
        current_turn_player_id=player_id,
    )
    game_sessions[game_id] = state
    return game_id, player_id


def add_player(game_id: str, player_name: str) -> str | None:
    """
    Add a player to a pending session. Returns new player_id, or None if the
    game is full / not in pending state.
    """
    state = game_sessions.get(game_id)
    if not state or state.status != "pending":
        return None

    if len(state.players) >= len(START_COUNTRIES):
        return None  # lobby full

    player_id = str(uuid.uuid4())
    start_country = START_COUNTRIES[len(state.players)]
    state.countries[start_country].owner_id = player_id

    player = Player(
        player_id=player_id,
        name=player_name,
        empire=[start_country],
    )
    state.players.append(player)
    return player_id


def get_session(game_id: str) -> GameState | None:
    return game_sessions.get(game_id)


def update_state(game_id: str, state: GameState) -> None:
    game_sessions[game_id] = state

File: backend/src/game/stateManager.ts
import { randomUUID } from "crypto";
import { Country, GameState, Player } from "../game/models";
import { mapData, startCountries } from "../game/mapData";

export const sessions = new Map<string, GameState>();

// Create a new game session. Returns { gameId, playerId }.
export function createSession(creatorName: string): {
  gameId: string;
  playerId: string;
} {
  const gameId = randomUUID();
  const playerId = randomUUID();

  const startCountry = startCountries[0];

  const countries: Record<string, Country> = Object.fromEntries(
    Object.entries(mapData).map(([name, data]) => [
      name,
      { name, ownerId: null, ...data },
    ]),
  );
  countries[startCountry].ownerId = playerId;

  const player: Player = {
    playerId: playerId,
    name: creatorName,
    empire: [startCountry],
    resources: [],
    armyStrength: 100,
    hasActedThisTurn: false,
  };

  const state: GameState = {
    gameId: gameId,
    creatorId: playerId,
    players: [player],
    countries,
    currentTurnPlayerId: playerId,
    turnNumber: 1,
    storyLog: [],
    status: "pending",
    winnerId: null,
  };

  sessions.set(gameId, state);
  return { gameId, playerId };
}
// Add a player to a pending session. Returns new playerId, or null if the
// game is full / not in pending state.
export function addPlayer(gameId: string, playerName: string): string | null {
  const state = sessions.get(gameId) as GameState | undefined;
  if (!state || state.status !== "pending") {
    return null; // game not found or not pending
  }

  if (state.players.length >= startCountries.length) {
    return null; // lobby full
  }

  const playerId = randomUUID();
  const startCountry = startCountries[state.players.length];
  state.countries[startCountry].ownerId = playerId;

  const player: Player = {
    playerId: playerId,
    name: playerName,
    empire: [startCountry],
    resources: [],
    armyStrength: 0,
    hasActedThisTurn: false,
  };
  state.players.push(player);
  return playerId;
}
export function getSession(gameId: string): GameState | undefined {
  return sessions.get(gameId) as GameState | undefined;
}
export function updateState(gameId: string, state: GameState): void {
  sessions.set(gameId, state);
}
